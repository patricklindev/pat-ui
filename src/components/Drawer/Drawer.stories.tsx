import React, { useState } from 'react';
import { Drawer } from './Drawer';
import Button from '.././Button/Button';

export default {
  title: 'Drawer',
  component: Drawer,
};
   
const drawerChildrenStyle: React.CSSProperties = {
  textAlign: 'left', marginLeft: '1rem'
};
    
export const persistentDrawer = () => {
  const [isOpen, setIsOpen] = useState<boolean>(false);

  const handleToggle = () => {
    setIsOpen(!isOpen);
  };
        
  return (
    <div
      style={{
        margin: '2.5rem auto 0 auto',
        width: '500px',
        lineHeight: '5rem',
        border: 'solid 1px #D3D3D3',
        borderRadius: '10px',
        textAlign: 'center',
      }}
    >
      <Drawer
        open={isOpen}
        variant={'persistent'}
        onToggleCallback={handleToggle}
      >
        <div style={drawerChildrenStyle}>
          <h5>Inbox</h5>
          <h5>All Mail </h5>
          <h5>Send Email</h5>
          <hr></hr>
          <h5>Spam</h5>
          <h5>Trash</h5>
        </div>
      </Drawer>
      <Button
        btnType="primary"
        disabled={false}
        onClick={handleToggle}
        style={{
          fontWeight: 'bold',
          lineHeight: '50px',
          alignSelf: 'center',
          padding: '0 10px',
        }}
      >
        Open/Close
      </Button>
    </div>
  );
};

const tempDrawerBtnStyle: React.CSSProperties = {
  fontWeight: 'bold',
  height: '50px',
  alignSelf: 'center',
  padding: '0 10px',
  width: '5rem',
  lineHeight: '50px',
  margin: '0 0.5rem'
};

export const temporaryDrawer = () => {
  const [isOpen, setIsOpen] = useState<boolean>(false);
  const [anchor, setAnchor] = useState<string>('left');
  
  const handleToggle = () => {
    setIsOpen(!isOpen);
  };
  const handleAnchor = (direction: string) => {
    setAnchor(direction);
  };

  return (
    <div style={{ marginTop: '2.5rem' }}>
      <Drawer
        open={isOpen}
        variant={'temporary'}
        onToggleCallback={handleToggle}
        anchor={anchor}
      >
        <div style={drawerChildrenStyle}>
          <h5>Inbox</h5>
          <h5>All Mail </h5>
          <h5>Send Email</h5>
          <hr></hr>
          <h5>Spam</h5>
          <h5>Trash</h5>
        </div>
      </Drawer>
      <div
        style={{
          margin: '0 auto',
          width: '500px',
          lineHeight: '5rem',
          border: 'solid 1px #D3D3D3',
          borderRadius: '10px',
          textAlign: 'center',
        }}
      >
        <Button
          btnType="primary"
          disabled={false}
          onClick={() => {
            handleToggle();
            handleAnchor('left');
          }}
          style={tempDrawerBtnStyle}
        >
          Left
        </Button>
        <Button
          btnType="primary"
          disabled={false}
          onClick={() => {
            handleToggle();
            handleAnchor('right');
          }}
          style={tempDrawerBtnStyle}
        >
          Right
        </Button>
        <Button
          btnType="primary"
          disabled={false}
          onClick={() => {
            handleToggle();
            handleAnchor('top');
          }}
          style={tempDrawerBtnStyle}
        >
          Top
        </Button>
        <Button
          btnType="primary"
          disabled={false}
          onClick={() => {
            handleToggle();
            handleAnchor('bottom');
          }}
          style={tempDrawerBtnStyle}
        >
          Bottom
        </Button>
      </div>
    </div>
  );
};
