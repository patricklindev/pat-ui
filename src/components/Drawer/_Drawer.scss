* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

.drawer {
  background-color: rgb(18, 18, 18);
  color: rgb(255, 255, 255);
  transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  // box-shadow: rgb(0, 0, 0 / 20%) 0px 8px 10px -5px, rgb(0, 0, 0 / 14%) 0px 16px 24px 2px, rgb(0,0, 0 / 12%) 0px 6px 30px 5px;
  background-image: linear-gradient(rgba(255, 255, 255, 0.15), rgba(255, 255, 255, 0.15));
  height: 100vh;
  padding: 30px;
  z-index: 1200;

  position: fixed;

}

.drawer-openStateFalse {
  display: none;
}

.drawer-left-open {
  top: 0;
  left: 0;
  height: 100vh;
  width: 250px;
  animation: slideInLeft 0.7s forwards;
}

@keyframes slideInLeft {
  0% {
    transform: translateX(-250px);
  }
  100% {
    transform: translateX(0);
  }
}

.drawer-right-open {
  top: 0;
  right: 0;
  height: 100vh;
  width: 250px;
  animation: slideInRight 0.7s forwards;
}

@keyframes slideInRight {
  0% {
    transform: translateX(900px);
  }
  100% {
    transform: translateX(0);
  }
}

.drawer-top-open {
  top: 0;
  width: 100vw;
  height: max-content;
  animation: slideInTop 0.7s forwards;
}

@keyframes slideInTop {
  0% {
    transform: translateY(-900px);
  }
  100% {
    transform: translateY(0);
  }
}

.drawer-bottom-open {
  bottom: 0;
  width: 100vw;
  height: max-content;
  animation: slideInBottom 0.7s forwards;
}

@keyframes slideInBottom {
  0% {
    transform: translateY(900px);
  }
  100% {
    transform: translateY(0);
  }
}

.drawer-dimBackground {
  position: fixed;
  display: flex;
  -webkit-box-align: center;
  align-items: center;
  -webkit-box-pack: center;
  justify-content: center;
  inset: 0px;
  background-color: rgba(0, 0, 0, 0.5);
  -webkit-tap-highlight-color: transparent;
  z-index: 1000;
}
