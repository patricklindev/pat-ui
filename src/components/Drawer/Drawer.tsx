import React, { FC, useState, useEffect, useRef } from 'react';
import { classNames } from '../../utils/classNames';
import './_Drawer.scss';

type Variant =  'persistent' | 'temporary';

type Anchor = 'top' | 'left' | 'bottom' | 'right' 

interface IDrawerProps {
  /** Sets the Drawer's position */
  anchor?: Anchor;
  /** Allows developer to set the variant of the Drawer, ( persistent | temporary ) */
  variant?: Variant;
  /** Sets the Drawer's open or closed state */
  open?: boolean;
  /** The user must pass in a callback function for controling the open / close state of the drawer */
  onToggleCallback?: () => void;
  /** Allows developer to set custom styles to the Drawer */
  className?: string; 
}
 
/**     
 * A Drawer component is used to display information or a meun on the side of the page.
 *
 * ```js
 * import { Drawer } from 'pat-ui'
 * ```
 * 
 *  
 * 
 * 
 *  Example of a onToggleCallback function that controls the open state
 * 
 * ```js
 *  const [isOpen, setIsOpen] = useState<boolean>(true);
 * 
 *  function handleToggle (){
      setIsOpen(!isOpen);
 *  }
 * 
 * ```
 * 
 * ```html
 * <Drawer open={isOpen} onToggleCallback={handleToggle} />
 * ```
 * 
 */
 
const Drawer: FC<DrawerProps> = ({ 
  anchor = 'left',
  open = false,
  onToggleCallback,
  variant = 'temporary',
  children,
  className,
}): JSX.Element => {

 

  let styleClasses = classNames('drawer', {

    [`${open ? `drawer-${anchor}-open` : 'drawer-openStateFalse'}`]: true,

  });
  if (className) {
    styleClasses += ' ' + className;
  }


  //click on backdrop to close drawer feature

  const handleToggleDrawer = (event: React.MouseEvent) => {
    if (onToggleCallback) {
        onToggleCallback();
    }
};

  // dimmed background

  let dimBackgroundStyle = '';
  if (variant === 'temporary' && open) {
    dimBackgroundStyle = 'drawer-dimBackground';
  }


  return (
    <section>
      <div className={dimBackgroundStyle} onClick={handleToggleDrawer} data-testid='dimmed-background'></div>
      <aside className={styleClasses} data-testid="drawer">
        {children}
      </aside>
    </section>
  );
};

export {Drawer};

