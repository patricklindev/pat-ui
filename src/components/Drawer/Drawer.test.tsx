import React from 'react';

import {Drawer} from './Drawer';

import { fireEvent, render, screen } from '@testing-library/react';

describe('Drawer', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(<Drawer />);
    expect(asFragment()).toMatchSnapshot();
  });


  it('Developers can control the open state of the component from props.', () => {
    render(<Drawer open={true} />);
    const drawerElement = screen.getByTestId('drawer');
    expect(drawerElement).toHaveClass('drawer__open_left');
    expect(drawerElement).not.toHaveClass('drawer__open_false');
  });

  it('Developers can control the close state of the component from props.', () => {
    render(<Drawer open={false} />);
    const drawerElement = screen.getByTestId('drawer');
    expect(drawerElement).toHaveClass('drawer__open_false');
    expect(drawerElement).not.toHaveClass('drawer__open_left');
  });

  it('Developers should see a dimed backdrop when the drawer component is open when style is set to temporary.', () => {
    render(<Drawer open={true} variant="temporary" />);
    const drawerElement = screen.getByTestId('dimmed-background');
    expect(drawerElement).toHaveClass('drawer__temporary_dimBackground');
  });

  it('Developers should not see a dimed backdrop when the drawer component is closed.', () => {
    render(<Drawer open={false} variant="temporary" />);
    const dimmedBackgroundDivElement = screen.getByTestId('dimmed-background');
    expect(dimmedBackgroundDivElement).not.toHaveClass('drawer__temporary_dimBackground');
  });

  it('Developers should not see a dimed backdrop when the drawer component style is set to persistent.', () => {
    render(<Drawer open={true} variant="persistent" />);
    const dimmedBackgroundDivElement = screen.getByTestId('dimmed-background');
    expect(dimmedBackgroundDivElement).not.toHaveClass('drawer__temporary_dimBackground');
  });

  it('Developers should be able to choose the left open direction of the drawer component from props.', () => {
    render(<Drawer open={true} anchor="left" />);
    const drawerElement = screen.getByTestId('drawer');
    expect(drawerElement).toHaveClass('drawer__open_left');
    expect(drawerElement).not.toHaveClass('drawer__open_right');
    expect(drawerElement).not.toHaveClass('drawer__open_top');
    expect(drawerElement).not.toHaveClass('drawer__open_bottom');
  });

  it('Developers should be able to choose the right open direction of the drawer component from props.', () => {
    render(<Drawer open={true} anchor="right" />);
    const drawerElement = screen.getByTestId('drawer');
    expect(drawerElement).toHaveClass('drawer__open_right');
    expect(drawerElement).not.toHaveClass('drawer__open_left');
    expect(drawerElement).not.toHaveClass('drawer__open_top');
    expect(drawerElement).not.toHaveClass('drawer__open_bottom');
  });

  it('Developers should be able to choose the top open direction of the drawer component from props.', () => {
    render(<Drawer open={true} anchor="top" />);
    const drawerElement = screen.getByTestId('drawer');
    expect(drawerElement).toHaveClass('drawer__open_top');
    expect(drawerElement).not.toHaveClass('drawer__open_left');
    expect(drawerElement).not.toHaveClass('drawer__open_right');
    expect(drawerElement).not.toHaveClass('drawer__open_bottom');
  });

  it('Developers should be able to choose the bottom open direction of the drawer component from props.', () => {
    render(<Drawer open={true} anchor="bottom" />);
    const drawerElement = screen.getByTestId('drawer');
    expect(drawerElement).toHaveClass('drawer__open_bottom');
    expect(drawerElement).not.toHaveClass('drawer__open_left');
    expect(drawerElement).not.toHaveClass('drawer__open_right');
    expect(drawerElement).not.toHaveClass('drawer__open_top');
  });

  it('will invoke the handleToggleDrawer once the user clicks on the backdrop', () => {
    const mockHandleToggleDrawer = jest.fn();

    render(<Drawer open onToggleCallback={mockHandleToggleDrawer} />);

    expect(mockHandleToggleDrawer).toBeCalledTimes(0);

    const backdrop = screen.getByTestId('dimmed-background');

    fireEvent.click(backdrop);

    expect(mockHandleToggleDrawer).toBeCalledTimes(1);
  });
});
