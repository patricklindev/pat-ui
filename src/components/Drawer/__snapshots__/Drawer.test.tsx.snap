// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer should match snapshot 1`] = `
<DocumentFragment>
  <section>
    <div
      class=""
      data-testid="dimmed-background"
    />
    <aside
      class="drawer drawer__open_false"
      data-testid="drawer"
    />
  </section>
</DocumentFragment>
`;
